File: harness/e2e/c/combined_harness.c
#include <stdint.h>

#include "harness_abi.h"
#include "../../../c/tmr/tmr.h"
#include "../../../c/checker/checker.h"
#include "../../../c/checkpoint/checkpoint.h"
#include "../../../c/recovery_block/recovery_block.h"
#include "../../../c/control_flow/control_flow.h"

/* The combined harness runs one workflow that chains every technique:
 *
 *   start -> read_input -> compute -> validate -> commit -> done
 *
 *   read_input  TMR triplet, majority vote
 *   compute     recovery block (primary -> acceptance test -> alternate)
 *   validate    checker acceptance gate
 *   commit      checkpoint commit-or-restart
 *   whole run   control-flow signature monitor wraps every transition
 *
 * A single inject point (harness_injection_point_before_workflow) sets the
 * (target, value) pair; the workflow applies each fault at its natural phase,
 * exactly as the standalone harnesses do. The final outcome is classified into
 * HARNESS_OUTCOME_* and a pass means the protected workflow avoided silent data
 * corruption. The baseline harness runs the identical workload unprotected. */

_Static_assert((int)TMR_OK == HARNESS_STATUS_OK,
    "tmr_status_t::TMR_OK must match HARNESS_STATUS_OK");
_Static_assert((int)TMR_ERR_NO_MAJORITY == HARNESS_STATUS_NO_MAJORITY,
    "tmr_status_t::TMR_ERR_NO_MAJORITY must match HARNESS_STATUS_NO_MAJORITY");
_Static_assert((int)CHECKPOINT_RESTART_COMMITTED == HARNESS_RESTART_COMMITTED,
    "checkpoint_restart_status_t::CHECKPOINT_RESTART_COMMITTED must match HARNESS_RESTART_COMMITTED");
_Static_assert((int)CHECKPOINT_RESTART_RESTORED == HARNESS_RESTART_RESTORED,
    "checkpoint_restart_status_t::CHECKPOINT_RESTART_RESTORED must match HARNESS_RESTART_RESTORED");
_Static_assert((int)CHECKPOINT_RESTART_RESTORE_FAILED == HARNESS_RESTART_RESTORE_FAILED,
    "checkpoint_restart_status_t::CHECKPOINT_RESTART_RESTORE_FAILED must match HARNESS_RESTART_RESTORE_FAILED");
_Static_assert((int)RECOVERY_BLOCK_PRIMARY_ACCEPTED == HARNESS_RECOVERY_PRIMARY_ACCEPTED,
    "recovery_block_status_t::RECOVERY_BLOCK_PRIMARY_ACCEPTED must match HARNESS_RECOVERY_PRIMARY_ACCEPTED");
_Static_assert((int)RECOVERY_BLOCK_ALTERNATE_ACCEPTED == HARNESS_RECOVERY_ALTERNATE_ACCEPTED,
    "recovery_block_status_t::RECOVERY_BLOCK_ALTERNATE_ACCEPTED must match HARNESS_RECOVERY_ALTERNATE_ACCEPTED");
_Static_assert((int)CONTROL_FLOW_OK == HARNESS_CONTROL_OK,
    "control_flow_status_t::CONTROL_FLOW_OK must match HARNESS_CONTROL_OK");
_Static_assert((int)CONTROL_FLOW_ERR_INVALID_TRANSITION == HARNESS_CONTROL_INVALID_TRANSITION,
    "control_flow_status_t::CONTROL_FLOW_ERR_INVALID_TRANSITION must match HARNESS_CONTROL_INVALID_TRANSITION");
_Static_assert((int)CONTROL_FLOW_ERR_BAD_SIGNATURE == HARNESS_CONTROL_BAD_SIGNATURE,
    "control_flow_status_t::CONTROL_FLOW_ERR_BAD_SIGNATURE must match HARNESS_CONTROL_BAD_SIGNATURE");

volatile uint32_t harness_iteration;
volatile uint32_t harness_stage;
volatile uint32_t harness_fault_target;
volatile uint32_t harness_fault_value;
volatile uint32_t harness_last_fault_target;
volatile uint32_t harness_last_expected;
volatile uint32_t harness_last_value;
volatile uint32_t harness_last_outcome;
volatile uint32_t harness_last_tmr_status;
volatile uint32_t harness_last_recovery_status;
volatile uint32_t harness_last_restart_status;
volatile uint32_t harness_last_control_status;
volatile uint32_t harness_last_active_check;
volatile uint32_t harness_last_checkpoint_check;
volatile uint32_t harness_last_phase;
volatile uint32_t harness_last_transitions;
volatile uint32_t harness_passes;
volatile uint32_t harness_failures;
volatile tmr_int_t harness_c_combined_tmr;
volatile checkpoint_record_t harness_c_combined_record;

static uint32_t sample_input(uint32_t iteration) {
    return 100u + ((iteration * 41u) % 700u);
}

static checker_record_t sample_record(uint32_t value) {
    return checker_record_init(
        CHECKER_TAG_SAMPLE,
        value,
        0u,
        1000u,
        6u,
        16u);
}

__attribute__((used, noinline))
void harness_injection_point_before_workflow(void) {
    __asm__ volatile("nop");
}

__attribute__((used, noinline))
void harness_injection_point_after_workflow(void) {
    __asm__ volatile("nop");
}

static void mirror_monitor(const control_flow_monitor_t *monitor) {
    harness_last_phase = monitor->phase;
    harness_last_transitions = monitor->transitions;
}

/* read_input: corrupt the TMR triplet before the majority vote. */
static void apply_tmr_fault(tmr_int_t *triplet) {
    const int value = (int)harness_fault_value;

    switch (harness_last_fault_target) {
    case HARNESS_FAULT_COPY_A:
        tmr_int_inject_fault_a(triplet, value);
        break;
    case HARNESS_FAULT_ALL_DISTINCT:
        tmr_int_inject_all(
            triplet,
            value,
            value ^ 0x11111111,
            value ^ 0x22222222);
        break;
    default:
        break;
    }
}

/* compute: corrupt the primary result so the acceptance test rejects it and
 * the recovery block falls through to the alternate. */
static void apply_after_primary_fault(checkpoint_record_t *state, void *context) {
    (void)context;

    harness_stage = HARNESS_STAGE_AFTER_PRIMARY;

    switch (harness_last_fault_target) {
    case HARNESS_FAULT_RECOVERY_PRIMARY_VALUE:
        state->active.value = harness_fault_value;
        break;
    case HARNESS_FAULT_RECOVERY_PRIMARY_CHECKSUM:
        state->active.checksum ^= harness_fault_value;
        break;
    default:
        break;
    }
}

static void apply_after_alternate_fault(checkpoint_record_t *state, void *context) {
    (void)state;
    (void)context;

    harness_stage = HARNESS_STAGE_AFTER_ALTERNATE;
}

/* control: corrupt the monitor between read_input and compute. */
static void apply_control_fault(control_flow_monitor_t *monitor) {
    switch (harness_last_fault_target) {
    case HARNESS_FAULT_CONTROL_PHASE:
        monitor->phase = harness_fault_value;
        break;
    case HARNESS_FAULT_CONTROL_SIGNATURE:
        monitor->signature ^= harness_fault_value;
        break;
    default:
        break;
    }
    mirror_monitor(monitor);
}

/* commit: corrupt the active record after validate and before commit so the
 * checkpoint commit-or-restart detects it and restores the last good state. */
static void apply_commit_fault(checkpoint_record_t *state) {
    switch (harness_last_fault_target) {
    case HARNESS_FAULT_ACTIVE_VALUE:
        state->active.value = harness_fault_value;
        break;
    case HARNESS_FAULT_ACTIVE_CHECKSUM:
        state->active.checksum ^= harness_fault_value;
        break;
    default:
        break;
    }
}

static int record_control(control_flow_status_t status) {
    harness_last_control_status = (uint32_t)status;
    return control_flow_passed(status);
}

/* Runs the full protected workflow once and leaves the observed facts in the
 * harness_last_* globals. harness_last_outcome is set by classify_outcome. */
static void run_workflow(uint32_t iteration) {
    control_flow_monitor_t monitor = control_flow_monitor_init();
    checkpoint_record_t *record =
        (checkpoint_record_t *)&harness_c_combined_record;
    tmr_int_t *triplet = (tmr_int_t *)&harness_c_combined_tmr;
    recovery_block_sample_update_t update;
    recovery_block_result_t recovery;
    checkpoint_restart_result_t commit;
    control_flow_status_t cf;
    tmr_status_t tmr_status;
    int voted = 0;
    uint32_t sample;

    mirror_monitor(&monitor);

    /* start -> read_input */
    cf = control_flow_monitor_advance(
        &monitor, CONTROL_FLOW_PHASE_START, CONTROL_FLOW_PHASE_READ_INPUT);
    if (!record_control(cf)) {
        mirror_monitor(&monitor);
        return;
    }

    /* read_input: TMR vote (faults applied to the triplet first). */
    harness_stage = HARNESS_STAGE_AFTER_CONTROL_READ;
    apply_control_fault(&monitor);
    apply_tmr_fault(triplet);
    tmr_status = tmr_int_read(triplet, &voted);
    harness_last_tmr_status = (uint32_t)tmr_status;
    if (tmr_status != TMR_OK) {
        return; /* no majority -> safe stop */
    }
    sample = (uint32_t)voted;

    /* read_input -> compute */
    cf = control_flow_monitor_advance(
        &monitor, CONTROL_FLOW_PHASE_READ_INPUT, CONTROL_FLOW_PHASE_COMPUTE);
    if (!record_control(cf)) {
        mirror_monitor(&monitor);
        return;
    }

    /* compute: recovery block over the voted sample. */
    harness_stage = HARNESS_STAGE_AFTER_CONTROL_COMPUTE;
    update.sample = sample;
    update.faults = RECOVERY_BLOCK_SAMPLE_FAULT_NONE;
    recovery = recovery_block_run_with_hooks(
        record,
        recovery_block_sample_primary,
        apply_after_primary_fault,
        recovery_block_sample_alternate,
        apply_after_alternate_fault,
        &update);
    harness_last_recovery_status = recovery.status;
    harness_last_checkpoint_check = recovery.checkpoint_check;
    harness_last_active_check = recovery.primary_check;
    if (recovery.status != RECOVERY_BLOCK_PRIMARY_ACCEPTED &&
        recovery.status != RECOVERY_BLOCK_ALTERNATE_ACCEPTED) {
        harness_last_value = record->active.value;
        return; /* unrecoverable -> safe stop (restored to last good) */
    }

    /* compute -> validate */
    cf = control_flow_monitor_advance(
        &monitor, CONTROL_FLOW_PHASE_COMPUTE, CONTROL_FLOW_PHASE_VALIDATE);
    if (!record_control(cf)) {
        mirror_monitor(&monitor);
        return;
    }

    /* validate: checker acceptance gate. */
    harness_last_active_check = checker_record_validate(&record->active);

    /* validate -> commit */
    cf = control_flow_monitor_advance(
        &monitor, CONTROL_FLOW_PHASE_VALIDATE, CONTROL_FLOW_PHASE_COMMIT);
    if (!record_control(cf)) {
        mirror_monitor(&monitor);
        return;
    }

    /* commit: checkpoint commit-or-restart (faults applied just before). */
    apply_commit_fault(record);
    commit = checkpoint_record_commit_or_restart(record);
    harness_last_restart_status = commit.status;
    harness_last_active_check = commit.active_check;
    harness_last_checkpoint_check = commit.checkpoint_check;
    harness_last_value = record->active.value;
    if (commit.status == CHECKPOINT_RESTART_RESTORE_FAILED) {
        return; /* both copies bad -> safe stop */
    }

    /* commit -> done */
    cf = control_flow_monitor_advance(
        &monitor, CONTROL_FLOW_PHASE_COMMIT, CONTROL_FLOW_PHASE_DONE);
    if (!record_control(cf)) {
        mirror_monitor(&monitor);
        return;
    }

    if (!record_control(control_flow_monitor_finish(&monitor))) {
        mirror_monitor(&monitor);
        return;
    }
    mirror_monitor(&monitor);

    (void)iteration;
}

/* Folds the observed facts into a single workflow outcome. */
static uint32_t classify_outcome(uint32_t expected) {
    const uint32_t corrected =
        harness_last_recovery_status == HARNESS_RECOVERY_ALTERNATE_ACCEPTED ||
        harness_last_restart_status == HARNESS_RESTART_RESTORED ||
        harness_c_combined_tmr.fault_count != 0u;

    /* A detected error that stopped the workflow before commit is fail-safe. */
    if (harness_last_control_status != HARNESS_CONTROL_OK ||
        harness_last_tmr_status != HARNESS_STATUS_OK ||
        harness_last_restart_status == HARNESS_RESTART_RESTORE_FAILED ||
        (harness_last_recovery_status != HARNESS_RECOVERY_PRIMARY_ACCEPTED &&
         harness_last_recovery_status != HARNESS_RECOVERY_ALTERNATE_ACCEPTED)) {
        return HARNESS_OUTCOME_SAFE_STOP;
    }

    if (harness_last_value == expected) {
        return corrected ? HARNESS_OUTCOME_RECOVERED : HARNESS_OUTCOME_CORRECT;
    }
    return HARNESS_OUTCOME_SDC;
}

static void validate(void) {
    const uint32_t target = harness_last_fault_target;
    const uint32_t outcome = harness_last_outcome;
    int pass;

    if (target == HARNESS_FAULT_NONE) {
        pass = outcome == HARNESS_OUTCOME_CORRECT;
    } else {
        pass = outcome != HARNESS_OUTCOME_SDC;
    }

    if (pass) {
        harness_passes += 1u;
    } else {
        harness_failures += 1u;
    }
}

void harness_main(void) {
    harness_stage = HARNESS_STAGE_BOOT;

    for (;;) {
        const uint32_t iteration = harness_iteration + 1u;
        const uint32_t input = sample_input(iteration);
        const uint32_t expected = recovery_block_sample_primary_value(input);

        harness_iteration = iteration;
        harness_last_expected = expected;
        harness_last_value = 0u;
        harness_last_outcome = HARNESS_OUTCOME_CORRECT;
        harness_last_tmr_status = HARNESS_STATUS_OK;
        harness_last_recovery_status = HARNESS_RECOVERY_PRIMARY_ACCEPTED;
        harness_last_restart_status = HARNESS_RESTART_COMMITTED;
        harness_last_control_status = HARNESS_CONTROL_OK;
        harness_last_active_check = CHECKER_OK;
        harness_last_checkpoint_check = CHECKER_OK;
        harness_last_phase = CONTROL_FLOW_PHASE_START;
        harness_last_transitions = 0u;
        harness_last_fault_target = HARNESS_FAULT_NONE;
        harness_c_combined_tmr = tmr_int_init((int)input);
        harness_c_combined_record = checkpoint_record_init(sample_record(input));

        harness_stage = HARNESS_STAGE_BEFORE_WORKFLOW;
        harness_injection_point_before_workflow();

        harness_last_fault_target = harness_fault_target;
        run_workflow(iteration);
        harness_fault_target = HARNESS_FAULT_NONE;

        harness_last_outcome = classify_outcome(expected);

        harness_stage = HARNESS_STAGE_AFTER_WORKFLOW;
        validate();
        harness_injection_point_after_workflow();
    }
}

File: harness/e2e/zig/combined_harness.zig
const abi = @import("abi");
const tmr = @import("tmr");
const checker = @import("checker");
const checkpoint = @import("checkpoint");
const recovery_block = @import("recovery_block");
const control_flow = @import("control_flow");
const std = @import("std");

// The combined harness runs one workflow that chains every technique:
//
//   start -> read_input -> compute -> validate -> commit -> done
//
//   read_input  TMR triplet, majority vote
//   compute     recovery block (primary -> acceptance test -> alternate)
//   validate    checker acceptance gate
//   commit      checkpoint commit-or-restart
//   whole run   control-flow signature monitor wraps every transition
//
// A single inject point sets the (target, value) pair; the workflow applies
// each fault at its natural phase. The final outcome is classified into
// abi.outcome.* and a pass means the protected workflow avoided silent data
// corruption. The baseline harness runs the identical workload unprotected.

comptime {
    std.debug.assert(abi.status.ok == 0);
    std.debug.assert(abi.status.no_majority == 1);
    std.debug.assert(abi.recovery.primary_accepted == @intFromEnum(recovery_block.RecoveryStatus.primary_accepted));
    std.debug.assert(abi.recovery.alternate_accepted == @intFromEnum(recovery_block.RecoveryStatus.alternate_accepted));
    std.debug.assert(abi.restart.committed == @intFromEnum(checkpoint.RestartStatus.committed));
    std.debug.assert(abi.restart.restored == @intFromEnum(checkpoint.RestartStatus.restored));
    std.debug.assert(abi.restart.restore_failed == @intFromEnum(checkpoint.RestartStatus.restore_failed));
    std.debug.assert(abi.control.ok == @intFromEnum(control_flow.ControlStatus.ok));
    std.debug.assert(abi.control.invalid_transition == @intFromEnum(control_flow.ControlStatus.invalid_transition));
    std.debug.assert(abi.control.bad_signature == @intFromEnum(control_flow.ControlStatus.bad_signature));
}

const TmrU32 = tmr.Tmr(u32);

export var harness_iteration: u32 = 0;
export var harness_stage: u32 = 0;
export var harness_fault_target: u32 = 0;
export var harness_fault_value: u32 = 0;
export var harness_last_fault_target: u32 = 0;
export var harness_last_expected: u32 = 0;
export var harness_last_value: u32 = 0;
export var harness_last_outcome: u32 = 0;
export var harness_last_tmr_status: u32 = 0;
export var harness_last_recovery_status: u32 = 0;
export var harness_last_restart_status: u32 = 0;
export var harness_last_control_status: u32 = 0;
export var harness_last_active_check: u32 = 0;
export var harness_last_checkpoint_check: u32 = 0;
export var harness_last_phase: u32 = 0;
export var harness_last_transitions: u32 = 0;
export var harness_passes: u32 = 0;
export var harness_failures: u32 = 0;

// Working state for the TMR vote and the recovery/checkpoint record. Both are
// initialized before the injection point and operated on in place, so they are
// memory-resident and corruptible exactly like the C harness's
// `harness_combined_tmr` / `harness_combined_record`. The control-flow monitor
// stays a local (as in C) since it is created fresh inside the workflow.
export var harness_combined_tmr: TmrU32 = undefined;
export var harness_combined_record: checkpoint.CheckpointedRecord = undefined;

fn load(ptr: *const volatile u32) u32 {
    return ptr.*;
}

fn store(ptr: *volatile u32, value: u32) void {
    ptr.* = value;
}

fn sampleInput(iteration: u32) u32 {
    return 100 + ((iteration *% 41) % 700);
}

fn sampleRecord(value: u32) checker.CheckedRecord {
    return checker.CheckedRecord.init(.sample, value, 0, 1000, 6, 16);
}

export fn harness_injection_point_before_workflow() callconv(.c) void {
    asm volatile ("nop // injection_point_before_workflow");
}

export fn harness_injection_point_after_workflow() callconv(.c) void {
    asm volatile ("nop // injection_point_after_workflow");
}

fn mirrorMonitor(monitor: *const control_flow.Monitor) void {
    store(&harness_last_phase, monitor.phase);
    store(&harness_last_transitions, monitor.transitions);
}

// read_input: corrupt the TMR triplet before the majority vote.
fn applyTmrFault(triplet: *TmrU32) void {
    const value = load(&harness_fault_value);
    switch (load(&harness_last_fault_target)) {
        abi.fault.copy_a => triplet.injectFaultA(value),
        abi.fault.all_distinct => triplet.injectAll(
            value,
            value ^ 0x11111111,
            value ^ 0x22222222,
        ),
        else => {},
    }
}

// compute: corrupt the primary result so the acceptance test rejects it and
// the recovery block falls through to the alternate.
fn applyAfterPrimaryFault(
    state: *checkpoint.CheckpointedRecord,
    _: *const recovery_block.SampleUpdate,
) void {
    store(&harness_stage, abi.stage.after_primary);
    switch (load(&harness_last_fault_target)) {
        abi.fault.recovery_primary_value => state.active.value = load(&harness_fault_value),
        abi.fault.recovery_primary_checksum => state.active.checksum ^= load(&harness_fault_value),
        else => {},
    }
}

fn applyAfterAlternateFault(
    _: *checkpoint.CheckpointedRecord,
    _: *const recovery_block.SampleUpdate,
) void {
    store(&harness_stage, abi.stage.after_alternate);
}

// control: corrupt the monitor between read_input and compute.
fn applyControlFault(monitor: *control_flow.Monitor) void {
    switch (load(&harness_last_fault_target)) {
        abi.fault.control_phase => monitor.phase = load(&harness_fault_value),
        abi.fault.control_signature => monitor.signature ^= load(&harness_fault_value),
        else => {},
    }
    mirrorMonitor(monitor);
}

// commit: corrupt the active record after validate and before commit so the
// checkpoint commit-or-restart detects it and restores the last good state.
fn applyCommitFault(state: *checkpoint.CheckpointedRecord) void {
    switch (load(&harness_last_fault_target)) {
        abi.fault.active_value => state.active.value = load(&harness_fault_value),
        abi.fault.active_checksum => state.active.checksum ^= load(&harness_fault_value),
        else => {},
    }
}

fn recordControl(status: control_flow.ControlStatus) bool {
    store(&harness_last_control_status, status.code());
    return status.passed();
}

// Runs the full protected workflow once over the pre-initialized working state
// and leaves the observed facts in the harness_last_* globals. The outcome is
// computed by classifyOutcome.
fn runWorkflow() void {
    var monitor = control_flow.Monitor.init();
    var update = recovery_block.SampleUpdate{
        .sample = 0,
        .faults = recovery_block.sample_fault.none,
    };

    mirrorMonitor(&monitor);

    // start -> read_input
    if (!recordControl(monitor.advance(.start, .read_input))) {
        mirrorMonitor(&monitor);
        return;
    }

    // read_input: TMR vote (faults applied to the triplet first).
    store(&harness_stage, abi.stage.after_control_read);
    applyControlFault(&monitor);
    applyTmrFault(&harness_combined_tmr);
    const voted = harness_combined_tmr.read() catch {
        store(&harness_last_tmr_status, abi.status.no_majority);
        return; // no majority -> safe stop
    };
    store(&harness_last_tmr_status, abi.status.ok);
    update.sample = voted;

    // read_input -> compute
    if (!recordControl(monitor.advance(.read_input, .compute))) {
        mirrorMonitor(&monitor);
        return;
    }

    // compute: recovery block over the voted sample.
    store(&harness_stage, abi.stage.after_control_compute);
    const recovery = recovery_block.runWithHooks(
        &harness_combined_record,
        &update,
        recovery_block.samplePrimary,
        applyAfterPrimaryFault,
        recovery_block.sampleAlternate,
        applyAfterAlternateFault,
    );
    store(&harness_last_recovery_status, recovery.status.code());
    store(&harness_last_checkpoint_check, recovery.checkpoint_check.code());
    store(&harness_last_active_check, recovery.primary_check.code());
    if (recovery.status != .primary_accepted and recovery.status != .alternate_accepted) {
        store(&harness_last_value, harness_combined_record.active.value);
        return; // unrecoverable -> safe stop (restored to last good)
    }

    // compute -> validate
    if (!recordControl(monitor.advance(.compute, .validate))) {
        mirrorMonitor(&monitor);
        return;
    }

    // validate: checker acceptance gate.
    store(&harness_last_active_check, harness_combined_record.active.validate().code());

    // validate -> commit
    if (!recordControl(monitor.advance(.validate, .commit))) {
        mirrorMonitor(&monitor);
        return;
    }

    // commit: checkpoint commit-or-restart (faults applied just before).
    applyCommitFault(&harness_combined_record);
    const commit = harness_combined_record.commitOrRestart();
    store(&harness_last_restart_status, commit.status.code());
    store(&harness_last_active_check, commit.active_check.code());
    store(&harness_last_checkpoint_check, commit.checkpoint_check.code());
    store(&harness_last_value, harness_combined_record.active.value);
    if (commit.status == .restore_failed) {
        return; // both copies bad -> safe stop
    }

    // commit -> done
    if (!recordControl(monitor.advance(.commit, .done))) {
        mirrorMonitor(&monitor);
        return;
    }

    if (!recordControl(monitor.finish())) {
        mirrorMonitor(&monitor);
        return;
    }
    mirrorMonitor(&monitor);
}

// Folds the observed facts into a single workflow outcome.
fn classifyOutcome(expected: u32) u32 {
    const corrected =
        load(&harness_last_recovery_status) == abi.recovery.alternate_accepted or
        load(&harness_last_restart_status) == abi.restart.restored or
        harness_combined_tmr.fault_count != 0;

    const recovery_status = load(&harness_last_recovery_status);
    if (load(&harness_last_control_status) != abi.control.ok or
        load(&harness_last_tmr_status) != abi.status.ok or
        load(&harness_last_restart_status) == abi.restart.restore_failed or
        (recovery_status != abi.recovery.primary_accepted and
            recovery_status != abi.recovery.alternate_accepted))
    {
        return abi.outcome.safe_stop;
    }

    if (load(&harness_last_value) == expected) {
        return if (corrected) abi.outcome.recovered else abi.outcome.correct;
    }
    return abi.outcome.sdc;
}

fn validate() void {
    const target = load(&harness_last_fault_target);
    const outcome = load(&harness_last_outcome);
    const pass = if (target == abi.fault.none)
        outcome == abi.outcome.correct
    else
        outcome != abi.outcome.sdc;

    if (pass) {
        store(&harness_passes, load(&harness_passes) +% 1);
    } else {
        store(&harness_failures, load(&harness_failures) +% 1);
    }
}

export fn harness_main() callconv(.c) noreturn {
    store(&harness_stage, abi.stage.boot);

    while (true) {
        const iteration = load(&harness_iteration) +% 1;
        const input = sampleInput(iteration);
        const expected = recovery_block.samplePrimaryValue(input);

        store(&harness_iteration, iteration);
        store(&harness_last_expected, expected);
        store(&harness_last_value, 0);
        store(&harness_last_outcome, abi.outcome.correct);
        store(&harness_last_tmr_status, abi.status.ok);
        store(&harness_last_recovery_status, abi.recovery.primary_accepted);
        store(&harness_last_restart_status, abi.restart.committed);
        store(&harness_last_control_status, abi.control.ok);
        store(&harness_last_active_check, @intFromEnum(checker.CheckStatus.ok));
        store(&harness_last_checkpoint_check, @intFromEnum(checker.CheckStatus.ok));
        store(&harness_last_phase, @intFromEnum(control_flow.Phase.start));
        store(&harness_last_transitions, 0);
        store(&harness_last_fault_target, abi.fault.none);
        harness_combined_tmr = TmrU32.init(input);
        harness_combined_record = checkpoint.CheckpointedRecord.init(sampleRecord(input));

        store(&harness_stage, abi.stage.before_workflow);
        @call(.never_inline, harness_injection_point_before_workflow, .{});

        store(&harness_last_fault_target, load(&harness_fault_target));
        runWorkflow();
        store(&harness_fault_target, abi.fault.none);

        store(&harness_last_outcome, classifyOutcome(expected));

        store(&harness_stage, abi.stage.after_workflow);
        validate();
        @call(.never_inline, harness_injection_point_after_workflow, .{});
    }
}

pub fn panic(msg: []const u8, trace: ?*std.builtin.StackTrace, ra: ?usize) noreturn {
    _ = msg;
    _ = trace;
    _ = ra;
    while (true) {}
}
